inspector: split --cpu-prof-path to --cpu-prof-dir and --cpu-prof-name

joyeecheung opened the pull request:
To improve the integration of `--cpu-prof` with workers, this patch
splits `--cpu-prof-path` into `--cpu-prof-dir` and `--cpu-prof-name`,
so when a worker is launched from a thread that enables
`--cpu-prof`, if the parent thread sets `--cpu-prof-dir`, then the
profile of both thread would be generated to the specified directory.
If they end up specifying the same `--cpu-prof-name` the behavior
is undefined - the last profile will overwritten the first one.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
cc @nodejs/workers @nodejs/performance @nodejs/process

addaleax commented on src/env-inl.h:
```diff
@@ -678,6 +686,29 @@ inline void Environment::set_cpu_profile_path(const std::string& path) {
+  return prof_dir_;
+}
+
+inline void Environment::InitializeProfDir(const std::string& dir) {
```
We only call this once, maybe this is better to just put in `env.cc`?

mscdex commented:
Why `--prof-dir` instead of something more specific like `--cpu-prof-dir` ? Having non-V8 `--prof-*` options is confusing because we already have `--prof` and `--prof-process` for V8.

joyeecheung commented on src/node_options.cc:
```diff
@@ -149,6 +149,15 @@ void EnvironmentOptions::CheckOptions(std::vector<std::string>* errors) {
+    errors->push_back("--cpu-prof-name must be used with --cpu-prof");
+  }
+
+  // We may add support for other profiles in the future.
```
@mscdex See this line here, the intention is that --prof-dir can be used for other profiles in the future.

I believe it is possible to also write --prof output to a specified directory, it's just that v8 does not provide such an option at the moment, but we could work towards that. This flag is experimental anyway.

joyeecheung replied:
Maybe a more obvious name would be `--node-prof-dir`?

mscdex commented:
I prefer keeping the specific directory prefix in the flag names for clarity. If we support more profile types in the future it just gives more flexibility.

joyeecheung commented:
> I prefer keeping the specific directory prefix in the flag names for clarity. If we support more profile types in the future it just gives more flexibility.

How would that gives more flexibility? If the whole path is specified in the flag, how would `--cpu-prof-path` work with workers? With `--cpu-prof-path`, you either hard-code the entire path, or nothing at all. With `--prof-dir` and `--cpu-prof-name` you can only hard-code the directory but let Node.js generate the file name for you, which I believe is the more common use case.

joyeecheung commented:
Moved the directory initialization to env.cc and renamed `--prof-dir` to `--node-prof-dir` to avoid the ambiguity. @mcollina @addaleax @mscdex PTAL, thanks!

mscdex commented:
> With `--cpu-prof-path`, you either hard-code the entire path, or nothing at all.

I was referring to having individual profile type directory prefix flags like `--cpu-prof-dir` vs a `--prof-dir` for all profile types.

joyeecheung commented:
> I was referring to having individual profile type directory prefix flags like --cpu-prof-dir vs a --prof-dir for all profile types.

I see, that makes sense to me, I'll switch to that instead, thanks!

mscdex commented on src/env.cc:
```diff
@@ -845,6 +845,24 @@ void Environment::stop_sub_worker_contexts() {
 
+#if HAVE_INSPECTOR
+
+void Environment::InitializeNodeProfDir(const std::string& dir) {
```
s/InitializeNodeProfDir/InitializeCpuProfDir/ ?

joyeecheung replied:
Done!

joyeecheung commented:
If there are no more reviews, I'd like to land this later today.

joyeecheung commented:
Landed in 49d3d11ba7975b2c6df4ecab55b4619da46fcb95
